### PR TITLE
B1

### PR DESCRIPTION
**Pull request by gamakgrover2002**

### CODE REVIEW

File: index.html [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+    <meta name="viewport" content="width=device-width, initial-scale=1.0" />
+    <title>Product Fetching</title>
+    <link rel="stylesheet" href="./style.css"></link>
+  </head>
+  <body>
+    <div class="container">
+      <div class="heading">
+        <div class="logo">Logo</div>
+        <ul class="navs">
+          <li>Home</li>
```
Comment (thesaty4) [inline index.html]: use shared file, load using script.

File: index.html [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+          <li>Products</li>
+          <li>About</li>
+          <li>Contact</li>
+        </ul>
+        <div id="btn-group">
```
Comment (thesaty4) [inline index.html]: naming convention should be correct. BEM

File: script.js [hunk 3]
```diff
@@ -0,0 +1,141 @@
+class ProductAPI {
```
Comment (thesaty4) [inline script.js]: where is our BaseAPIService class, create BaseAPIService class and extend them in the your product service.

File: script.js [hunk 4]
```diff
@@ -0,0 +1,141 @@
...
+    try {
+      const response = await fetch(
+        `${this.baseURL}?limit=${limit}&skip=${offset}&select=title,price`
+      );
+      if (!response.ok) {
+        throw new Error(`Error fetching products: ${response.statusText}`);
+      }
+      const data = await response.json();
+      return data;
+    } catch (error) {
```
Comment (thesaty4) [inline script.js]: DRY ?????

File: script.js [hunk 5]
```diff
@@ -0,0 +1,141 @@
...
+      console.error("Error fetching product:", error);
+      throw error;
+    }
+  }
+}
+
+const baseURL = "https://dummyjson.com/products";
```
Comment (thesaty4) [inline script.js]: move every dependent variable in the ProductClass

File: script.js [hunk 6]
```diff
@@ -0,0 +1,141 @@
...
+const productAPI = new ProductAPI(baseURL);
+let limit = 9;
+let data = [];
+let offset = 0;
+let productDetails = {};
+let currentPage = 1;
+
+async function fetchData() {
+  try {
```
Comment (thesaty4) [inline script.js]: fetchData will part of the Product

File: script.js [hunk 7]
```diff
@@ -0,0 +1,141 @@
...
+    const result = await productAPI.apiGet(limit, offset);
+    data = result.products;
+    window.scrollTo(0, 0);
+    renderProducts();
+  } catch (error) {
+    console.error(error);
+  }
+}
+
+async function getProduct(id) {
```
Comment (thesaty4) [inline script.js]: same as above

File: script.js [hunk 8]
```diff
@@ -0,0 +1,141 @@
...
+  try {
+    const result = await productAPI.getProduct(id);
+    productDetails[id] = result;
+    renderProductDetails(id);
+  } catch (error) {
+    console.error(error);
+  }
+}
+
+function renderProducts() {
```
Comment (thesaty4) [inline script.js]: create seperated class to handle everything

File: script.js [hunk 9]
```diff
@@ -0,0 +1,141 @@
...
+    ${e.title}<br />
+    Price: $${e.price}<br />
+    <button onclick="getProduct(${e.id})">Get Details</button>
+    <div id="details-${e.id}"></div>
+  `;
+    productsContainer.appendChild(productDiv);
+  });
+}
+
+function renderProductDetails(id) {
```
Comment (thesaty4) [inline script.js]: method will be the part of the your class

File: script.js [hunk 10]
```diff
@@ -0,0 +1,141 @@
...
+  const detailsDiv = document.getElementById(`details-${id}`);
+  if (detailsDiv) {
+    detailsDiv.innerHTML = `
+    <p class="details">${productDetails[id].description}</p>
+    <p class="rating">${productDetails[id].rating}/5</p>
+  `;
+  }
+}
+
+function handleScroll() {
```
Comment (thesaty4) [inline script.js]: same as above

File: script.js [hunk 11]
```diff
@@ -0,0 +1,141 @@
...
+  const currentHeight = document.documentElement.scrollTop + window.innerHeight;
+  if (currentHeight - 20 > scrollHeight) {
+    currentPage += 1;
+    offset += limit;
+    fetchData();
+    renderPagination(currentPage, 22);
+  }
+}
+
+function renderPagination(currentPage, totalPages) {
```
Comment (thesaty4) [inline script.js]: samve as above

File: script.js [hunk 12]
```diff
@@ -0,0 +1,141 @@
...
+  }
+
+  pagination.innerHTML += `
+  <button class="side-button" onclick="handlePageChange(${currentPage + 1})" ${
+    currentPage === totalPages ? "disabled" : ""
+  }>Next</button>
+`;
+}
+
+function handlePageChange(pageNum) {
```
Comment (thesaty4) [inline script.js]: product class handle

File: style.css [hunk 13]
```diff
@@ -0,0 +1,82 @@
...
+#products-container {
+  display: grid;
+  grid-template-columns: auto auto auto;
+}
+.details {
+  font-size: 1vw;
+  margin-top: 3vh;
+}
+.logo {
+  width: 10vw;
```
Comment (thesaty4) [inline style.css]: logo should be fixed size

**Review by thesaty4 — CHANGES_REQUESTED** [review]
Review Points:

- Need to focus on improving UI/UX.
- Always think inside the box, not outside the box.
- Use class-based implementation for the APIs; OOP should be utilized.
- Implement the DOM with class-based methods.
- Offset for dropdown (10, 20, 50, 100).
- On hitting the bottom, the scroller should not go to the top.
- All data should exist.
- API caching (think like DP).
- Previous/Next: if data is already loaded, the API will not be called.
- On clicking any specific page: if data is already loaded, the API will not be called.

File: index.html (hunk 1 above)
Comment (gamakgrover2002) [reply index.html]: done

File: index.html (hunk 2 above)
Comment (gamakgrover2002) [reply index.html]: done

File: style.css (hunk 13 above)
Comment (gamakgrover2002) [reply style.css]: done

File: script.js (hunk 11 above)
Comment (gamakgrover2002) [reply script.js]: done

File: script.js (hunk 3 above)
Comment (gamakgrover2002) [reply script.js]: done

File: script.js (hunk 4 above)
Comment (gamakgrover2002) [reply script.js]: done

File: script.js (hunk 6 above)
Comment (gamakgrover2002) [reply script.js]: done

File: script.js (hunk 5 above)
Comment (gamakgrover2002) [reply script.js]: done

File: script.js (hunk 7 above)
Comment (gamakgrover2002) [reply script.js]: done

File: script.js (hunk 8 above)
Comment (gamakgrover2002) [reply script.js]: done

File: script.js (hunk 9 above)
Comment (gamakgrover2002) [reply script.js]: done

File: script.js (hunk 10 above)
Comment (gamakgrover2002) [reply script.js]: done

File: script.js (hunk 12 above)
Comment (gamakgrover2002) [reply script.js]: done

**Pull request closed without merging**